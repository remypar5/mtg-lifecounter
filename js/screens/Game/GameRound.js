import React from 'react';
import PropTypes from 'prop-types';
import { View, StyleSheet, Alert, BackHandler } from 'react-native';
import { addNavigationHelpers, NavigationActions } from 'react-navigation';

import Player from '../../components/Player';
import splitArray from '../../utils/array.split';

export default class GameRound extends React.Component {

    confirmGameOver = true;

    constructor(props) {
        super(props);

        this.players = this.generatePlayers();

        this.state = {
            roundNumber: 1
        };
    }

    componentDidMount() {
        BackHandler.addEventListener('hardwareBackPress', this.onBackPress);
    }

    componentWillUnmount() {
        BackHandler.removeEventListener('hardwareBackPress', this.onBackPress);
    }

    render() {
        const { length } = this.players;
        const { roundNumber } = this.state;

        return (
            <View style={ styles.container } key={ `round${roundNumber}` }>
                { length === 1 ?
                    this.renderSinglePlayer(this.players[0]) :
                    ( length === 2 ?
                        this.renderTwoPlayers() :
                        this.renderPlayers()) }
            </View>
        );
    }

    renderSinglePlayer(player, playerProps = {}) {
        return (
            <Player
                { ...playerProps }
                player={player}
                onGameOver={ (isGameOver) => this.playerGameOver(player, isGameOver) } />
        );
    }

    renderTwoPlayers() {
        const [ player1, player2 ] = this.players;
        const { asColumn, rotated180 } = styles;

        return (
            <View style={ asColumn }>
                <View style={[ asColumn, rotated180 ]}>
                    { this.renderSinglePlayer(player2) }
                </View>
                <View style={[asColumn]}>
                    { this.renderSinglePlayer(player1) }
                </View>
            </View>
        );
    }

    renderPlayers() {
        const columns = splitInHalf(this.players);
        const { column: columnStyle } = styles;

        return columns.map((players, column) => (
            <View key={`playerColumn${column}`} style={ columnStyle }>
                { players.map((player) => this.renderSinglePlayer(player, {
                    key: `player${player.id}`
                })) }
            </View>
        ));
    }

    generatePlayers() {
        const colors = ['#57719e', '#d83f2d', '#95b54c', '#efd86f', '#b2b2b0', '#cb8034'];
        const players = [];
        const { numberOfPlayers, startingLifeTotal } = this.props;

        for (let id = 0; id < numberOfPlayers; id++) {
            players.push({
                id,
                name: 'Player ' + (id+1),
                life: startingLifeTotal,
                isGameOver: false,
                color: colors[id]
            });
        }

        return players;
    }

    playerGameOver(player, isGameOver) {
        const self = this;
        const { numberOfPlayers, navigation } = this.props;
        const { goBack } = navigation;
        const minDeadPlayers = this.players.length === 1 ? 0 : 1;
        let gameOverPlayers = 0;

        player.isGameOver = isGameOver;

        for (let p of this.players) {
            if (p.isGameOver) {
                gameOverPlayers++;
            }
        }

        if (this.confirmGameOver && (numberOfPlayers - gameOverPlayers) <= minDeadPlayers) {
            Alert.alert('Game Over', 'What do you want to do?', [
                {
                    text: 'Continue game', onPress: () => {
                        self.confirmGameOver = false;
                        Alert.alert('Continue game', 'Press the back button on your device to exit the game.');
                    }
                },
                { text: 'Exit', onPress: () => goBack() },
                { text: 'Restart', onPress: () => self.setState({ roundNumber: this.state.roundNumber + 1 }) },
            ], { cancelable: true });
        }
    }

    onBackPress = () => {
        const { index, dispatch } = this.props.navigation;
        if (index === 0) {
            return false;
        }

        Alert.alert('Exit game', 'Are you sure?', [
            { text: 'Yes', onPress: () => dispatch(NavigationActions.back()) },
            { text: 'No' }
        ]);

        return true;
    }
}

GameRound.propsTypes = {
    numberOfPlayers: PropTypes.number.isRequired,
    startingLifeTotal: PropTypes.number.isRequired,
};

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#161616',
        flex: 1,
        flexDirection: 'row',
    },
    column: {
        width: '50%',
    },
    asColumn: {
        flex: 1,
        flexDirection: 'column',
    },
    rotated180: {
        transform: [
            { rotate: '180deg' },
        ]
    },
});

const splitInHalf = (arr) => {
    const half = Math.ceil(arr.length / 2);

    return splitArray(arr, [(item, idx) => idx < half, (item, idx) => idx >= half], true);
};
