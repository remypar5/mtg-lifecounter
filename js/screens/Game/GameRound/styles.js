import { StyleSheet } from 'react-native';

const styles = StyleSheet.create({
    container: {
        backgroundColor: '#161616',
        flex: 1,
        flexDirection: 'row',
    },
    column: {
        width: '50%',
    },
    asColumn: {
        flex: 1,
        flexDirection: 'column',
    },
    rotated180: {
        transform: [
            { rotate: '180deg' },
        ],
    },
});

export default styles;
