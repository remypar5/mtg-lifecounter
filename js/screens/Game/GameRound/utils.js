import arraySplit from '../../../utils';

const splitInHalf = (arr) => {
    const half = Math.ceil(arr.length / 2);

    return arraySplit(arr, [(item, idx) => idx < half, (item, idx) => idx >= half], true);
};

export default splitInHalf;
