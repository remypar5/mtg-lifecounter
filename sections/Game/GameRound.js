import React from 'react';
import PropTypes from 'prop-types';
import { Text, View, StyleSheet, Dimensions } from 'react-native';

import Player from '../../components/Player';
import splitArray from '../../utils/array.split';

export default class GameRound extends React.Component {
    constructor(props) {
        super(props);

        this.players = this.generatePlayers();
    }

    // TODO: Lock screen's orientation to portrait after ejecting
    // componentDidMount() {
    //     Orientation.lockToPortrait();
    // }

    render() {
        return (
            <View style={ styles.container }>
                { this.renderPlayers() }
            </View>
        );
    }

    renderPlayers() {
        const sides = ['first', 'second'];
        const columns = splitInHalf(this.players);

        return columns.map((players, column) => (
            <View key={`playerColumn${column}`} style={ [styles.column, styles[sides[column] + 'Column']] }>
                { players.map((player) => (
                    <Player key={`playerContainer${player.id}`} player={player} style={ styles[sides[column] + 'Player'] } />
                )) }
            </View>
        ));
    }
    
    generatePlayers() {
        const players = [];
        const props = this.props;
        
        for (let id = 0; id < props.numberOfPlayers; id++) {
            players.push({
                id,
                name: 'Player ' + (id+1),
                life: props.startingLifeTotal
            });
        }
        
        return players;
    }
}

GameRound.propsTypes = {
    numberOfPlayers: PropTypes.number.isRequired,
    startingLifeTotal: PropTypes.number.isRequired
};

const styles = StyleSheet.create({
    container: {
        flex: 1,
        flexDirection: 'row',
    },
    column: {
        width: '50%',
    },
    firstColumn: {
        //
    },
    secondColumn: {
        //
    },
    firstPlayer: {
        //
    },
    secondPlayer: {
        //
    }
});

const splitInHalf = (arr) => {
    const half = Math.ceil(arr.length / 2);
    
    return splitArray(arr, [(item, idx) => idx < half, (item, idx) => idx >= half], true)
};
