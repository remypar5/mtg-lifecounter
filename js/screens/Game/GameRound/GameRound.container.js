import React from 'react';
import PropTypes from 'prop-types';

import GameRound from './GameRound';

// Ignore the next line because react-router requires containers to be actual React Components
// eslint-disable-next-line react/prefer-stateless-function
export default class GameRoundContainer extends React.Component {
    render() {
        const { navigation } = this.props;
        const { numberOfPlayers, startingLifeTotal, stayAwake } = navigation.state.params;

        return (<GameRound
            navigation={navigation}
            numberOfPlayers={numberOfPlayers}
            startingLifeTotal={startingLifeTotal}
            stayAwake={stayAwake}
        />);
    }
}

GameRoundContainer.navigationOptions = {
    header: null,
};

GameRoundContainer.propTypes = {
    navigation: PropTypes.object.isRequired,
};
