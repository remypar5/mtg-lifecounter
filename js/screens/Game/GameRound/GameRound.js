import React from 'react';
import PropTypes from 'prop-types';
import { View, Alert, BackHandler } from 'react-native';
import { NavigationActions } from 'react-navigation';

import styles from './styles';
import splitInHalf from './utils';
import { Player, PageContainer } from '../../../components';

export default class GameRound extends React.Component {
    constructor(props) {
        super(props);

        this.confirmGameOver = true;
        this.players = this.generatePlayers();

        this.state = {
            roundNumber: 1,
        };

        this.onBackPress = this.onBackPress.bind(this);
    }

    componentDidMount() {
        BackHandler.addEventListener('hardwareBackPress', this.onBackPress);
    }

    componentWillUnmount() {
        BackHandler.removeEventListener('hardwareBackPress', this.onBackPress);
    }

    onBackPress() {
        const { index, dispatch } = this.props.navigation;
        if (index === 0) {
            return false;
        }

        Alert.alert('Exit game', 'Are you sure?', [
            { text: 'Yes', onPress: () => dispatch(NavigationActions.back()) },
            { text: 'No' },
        ]);

        return true;
    }

    generatePlayers() {
        const colors = ['#57719e', '#d83f2d', '#95b54c', '#efd86f', '#b2b2b0', '#cb8034'];
        const players = [];
        const { numberOfPlayers, startingLifeTotal } = this.props;

        for (let id = 0; id < numberOfPlayers; id++) {
            players.push({
                id,
                name: `Player ${id + 1}`,
                life: startingLifeTotal,
                isGameOver: false,
                color: colors[id],
            });
        }

        return players;
    }

    restart() {
        this.players = this.generatePlayers();

        const { roundNumber } = this.state;
        this.setState({
            roundNumber: roundNumber + 1,
        });
    }

    playerGameOver(player, isGameOver) {
        const self = this;
        const { numberOfPlayers, navigation } = this.props;
        const { goBack } = navigation;
        const minDeadPlayers = this.players.length === 1 ? 0 : 1;
        let gameOverPlayers = 0;

        // eslint-disable-next-line no-param-reassign
        player.isGameOver = isGameOver;

        this.players.forEach((p) => {
            if (p.isGameOver) {
                gameOverPlayers++;
            }
        });

        if (this.confirmGameOver && (numberOfPlayers - gameOverPlayers) <= minDeadPlayers) {
            Alert.alert('Game Over', 'What do you want to do?', [
                {
                    text: 'Continue game',
                    onPress: () => {
                        self.confirmGameOver = false;
                        Alert.alert('Continue game', 'Press the back button on your device to exit the game.');
                    },
                },
                { text: 'Exit', onPress: () => goBack() },
                { text: 'Restart', onPress: () => self.restart() },
            ], { cancelable: true });
        }
    }

    renderSinglePlayer(player, playerProps = {}) {
        const props = {
            size: this.players.length > 2 ? 'small' : 'large',
            ...playerProps,
            player,
        };

        return (
            <Player
                {...props}
                onGameOver={(isGameOver) => this.playerGameOver(player, isGameOver)}
            />
        );
    }

    renderTwoPlayers() {
        const [player1, player2] = this.players;
        const { asColumn, rotated180 } = styles;

        return (
            <View style={asColumn}>
                <View style={[asColumn, rotated180]}>
                    {this.renderSinglePlayer(player2)}
                </View>
                <View style={[asColumn]}>
                    {this.renderSinglePlayer(player1)}
                </View>
            </View>
        );
    }

    renderMultiPlayer() {
        const sides = ['Left', 'Right'];
        const columns = splitInHalf(this.players);
        const { column: columnStyle } = styles;

        return columns.map((players, column) => (
            <View key={`playerColumn${sides[column]}`} style={columnStyle}>
                { players.map((player) => this.renderSinglePlayer(player, {
                    key: `player${player.id}`,
                })) }
            </View>
        ));
    }

    renderPlayers() {
        const { length } = this.players;

        return length === 2 ? this.renderTwoPlayers() : this.renderMultiPlayer();
    }

    render() {
        const { length } = this.players;
        const { roundNumber } = this.state;

        return (
            <PageContainer key={`round${roundNumber}`}>
                { length === 1 ? this.renderSinglePlayer(this.players[0]) : this.renderPlayers() }
            </PageContainer>
        );
    }
}

GameRound.propTypes = {
    numberOfPlayers: PropTypes.number.isRequired,
    startingLifeTotal: PropTypes.number.isRequired,
    navigation: PropTypes.shape({
        dispatch: PropTypes.func.isRequired,
        index: PropTypes.number,
    }).isRequired,
};
