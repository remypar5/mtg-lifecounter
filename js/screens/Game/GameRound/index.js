import GameRound from './GameRound';
import GameRoundContainer from './GameRound.container';
import styles from './styles';

export {
    GameRound,
    GameRoundContainer,
    styles,
};
