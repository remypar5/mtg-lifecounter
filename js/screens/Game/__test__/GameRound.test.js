import React from 'react';
import { shallow } from 'enzyme';

import GameRound from '../GameRound';

describe('<GameRound />', () => {

    let component;

    beforeEach(() => {
        component = shallow(
            <GameRound
                numberOfPlayers={3}
                startingLifeTotal={40} />
        );
    });

    it('renders without crashing', () => {
        expect(component).toMatchSnapshot();
    });

});
