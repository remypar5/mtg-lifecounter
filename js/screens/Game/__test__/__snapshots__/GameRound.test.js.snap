// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GameRound /> renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GameRound
    numberOfPlayers={3}
    onGameEnd={[Function]}
    startingLifeTotal={40}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": "round1",
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <View
          style={
            Array [
              Object {
                "width": "50%",
              },
              Object {},
            ]
          }
        >
          <Player
            onGameOver={[Function]}
            player={
              Object {
                "id": 0,
                "isGameOver": false,
                "life": 40,
                "name": "Player 1",
              }
            }
            style={Object {}}
          />
          <Player
            onGameOver={[Function]}
            player={
              Object {
                "id": 1,
                "isGameOver": false,
                "life": 40,
                "name": "Player 2",
              }
            }
            style={Object {}}
          />
        </View>,
        <View
          style={
            Array [
              Object {
                "width": "50%",
              },
              Object {},
            ]
          }
        >
          <Player
            onGameOver={[Function]}
            player={
              Object {
                "id": 2,
                "isGameOver": false,
                "life": 40,
                "name": "Player 3",
              }
            }
            style={Object {}}
          />
        </View>,
      ],
      "style": Object {
        "backgroundColor": "black",
        "flex": 1,
        "flexDirection": "row",
      },
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "playerColumn0",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 0,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 1",
                }
              }
              style={Object {}}
            />,
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 1,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 2",
                }
              }
              style={Object {}}
            />,
          ],
          "style": Array [
            Object {
              "width": "50%",
            },
            Object {},
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "playerContainer0",
            "nodeType": "class",
            "props": Object {
              "onGameOver": [Function],
              "player": Object {
                "id": 0,
                "isGameOver": false,
                "life": 40,
                "name": "Player 1",
              },
              "style": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "playerContainer1",
            "nodeType": "class",
            "props": Object {
              "onGameOver": [Function],
              "player": Object {
                "id": 1,
                "isGameOver": false,
                "life": 40,
                "name": "Player 2",
              },
              "style": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "playerColumn1",
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 2,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 3",
                }
              }
              style={Object {}}
            />,
          ],
          "style": Array [
            Object {
              "width": "50%",
            },
            Object {},
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "playerContainer2",
            "nodeType": "class",
            "props": Object {
              "onGameOver": [Function],
              "player": Object {
                "id": 2,
                "isGameOver": false,
                "life": 40,
                "name": "Player 3",
              },
              "style": Object {},
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": "round1",
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <View
            style={
              Array [
                Object {
                  "width": "50%",
                },
                Object {},
              ]
            }
          >
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 0,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 1",
                }
              }
              style={Object {}}
            />
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 1,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 2",
                }
              }
              style={Object {}}
            />
          </View>,
          <View
            style={
              Array [
                Object {
                  "width": "50%",
                },
                Object {},
              ]
            }
          >
            <Player
              onGameOver={[Function]}
              player={
                Object {
                  "id": 2,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 3",
                }
              }
              style={Object {}}
            />
          </View>,
        ],
        "style": Object {
          "backgroundColor": "black",
          "flex": 1,
          "flexDirection": "row",
        },
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "playerColumn0",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Player
                onGameOver={[Function]}
                player={
                  Object {
                    "id": 0,
                    "isGameOver": false,
                    "life": 40,
                    "name": "Player 1",
                  }
                }
                style={Object {}}
              />,
              <Player
                onGameOver={[Function]}
                player={
                  Object {
                    "id": 1,
                    "isGameOver": false,
                    "life": 40,
                    "name": "Player 2",
                  }
                }
                style={Object {}}
              />,
            ],
            "style": Array [
              Object {
                "width": "50%",
              },
              Object {},
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "playerContainer0",
              "nodeType": "class",
              "props": Object {
                "onGameOver": [Function],
                "player": Object {
                  "id": 0,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 1",
                },
                "style": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "playerContainer1",
              "nodeType": "class",
              "props": Object {
                "onGameOver": [Function],
                "player": Object {
                  "id": 1,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 2",
                },
                "style": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "playerColumn1",
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Player
                onGameOver={[Function]}
                player={
                  Object {
                    "id": 2,
                    "isGameOver": false,
                    "life": 40,
                    "name": "Player 3",
                  }
                }
                style={Object {}}
              />,
            ],
            "style": Array [
              Object {
                "width": "50%",
              },
              Object {},
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "playerContainer2",
              "nodeType": "class",
              "props": Object {
                "onGameOver": [Function],
                "player": Object {
                  "id": 2,
                  "isGameOver": false,
                  "life": 40,
                  "name": "Player 3",
                },
                "style": Object {},
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
